cub/cub/device/device_reduce.cuh

```diff
@@ -473,6 +495,145 @@ struct DeviceReduce
+  //!
+  //! @endrst
+  //!
+  //! @tparam InputIteratorT
+  //!   **[inferred]** Random-access input iterator type for reading input items
+  //!   (of some type `T`) @iterator
+  //!
+  //! @tparam ExtremumOutIteratorT
+  //!   **[inferred]** Output iterator type for recording minimum value
+  //!
+  //! @tparam IndexOutIteratorT
+  //!   **[inferred]** Output iterator type for recording index of the returned value
+  //!
+  //! @param[in] d_temp_storage
+  //!   Device-accessible allocation of temporary storage. When `nullptr`, the
+  //!   required allocation size is written to `temp_storage_bytes` and no work is done.
+  //!
+  //! @param[in,out] temp_storage_bytes
+  //!   Reference to size in bytes of `d_temp_storage` allocation
+  //!
+  //! @param[in] d_in
+  //!   Iterator to the input sequence of data items
+  //!
+  //! @param[out] d_min_out
+  //!   Iterator to which the minimum value is written
+  //!
+  //! @param[out] d_index_out
+  //!   Iterator to which the index of the returned value is written
+  //!
+  //! @param[in] num_items
+  //!   Total number of input items (i.e., length of `d_in`)
+  //!
+  //! @param[in] stream
+  //!   @rst
+  //!   **[optional]** CUDA stream to launch kernels within. Default is stream\ :sub:`0`.
+  //!   @endrst
+  template <typename InputIteratorT, typename ExtremumOutIteratorT, typename IndexOutIteratorT>
+  CUB_RUNTIME_FUNCTION static cudaError_t ArgMin(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    InputIteratorT d_in,
+    ExtremumOutIteratorT d_min_out,
+    IndexOutIteratorT d_index_out,
+    ::cuda::std::int64_t num_items,
+    cudaStream_t stream = 0)
+  {
+    CUB_DETAIL_NVTX_RANGE_SCOPE_IF(d_temp_storage, "cub::DeviceReduce::ArgMin");
+
+    // Signed integer type for global offsets
+    using OffsetT = ::cuda::std::int64_t;
+
+    // The input type
+    using InputValueT = cub::detail::value_t<InputIteratorT>;
+
+    // The value type used for the extremum
+    using OutputExtremumT = detail::non_void_value_t<ExtremumOutIteratorT, InputValueT>;
+
+    // The (index, extremum)-tuple type
+    using OutputTupleT = KeyValuePair<OffsetT, OutputExtremumT>;
+
+    // Initial value type
+    using InitT = detail::reduce::empty_problem_init_t<OutputTupleT>;
+
+    // Accumulator type
+    using AccumT = OutputTupleT;
+
+    // Wrapped input iterator to produce index-value <OffsetT, InputT> tuples
+    using ArgIndexInputIteratorT = ArgIndexInputIterator<InputIteratorT, OffsetT, OutputExtremumT>;
+    ArgIndexInputIteratorT d_indexed_in(d_in);
+
+    // Initial value
+    InitT initial_value{AccumT(1, ::cuda::std::numeric_limits<InputValueT>::max())};
+
+    // Tabulate output iterator that unzips the result and writes it to the user-provided output iterators
+    auto unzip_result_it = THRUST_NS_QUALIFIER::make_tabulate_output_iterator(
+      detail::reduce::unzip_and_write_arg_extremum_op<ExtremumOutIteratorT, IndexOutIteratorT>{d_min_out, d_index_out});
```

Comment:
Suggestion: I would have written this differently: I would have pulled the separation of extremum and index deep into the dispatcher and avoid splitting the key value pair here. There is no technical reason for this yet, but my gut feeling tells me the abstraction depth (the amount of source code constructs we use that the compiler has to inline and optimize out) of this implementation seems higher than if we would pass two iterators for output to the dispatcher.

However, this can be refactored later.

Reply:
Ok sorry, I had a misunderstanding that @elstehle could clarify offline with me. I thougth that `ArgMin|Max` would use a different dispatcher than `Reduce`. However, given they use the same code path, we have to retain the key-value pair for index and value and leave the dispatcher as is.

Then my only (tiny) concern is that the tabulate output iterator carries a needless integer around. However, I assume the compiler could optimize it out. Would be interesting to see a SASS comparison between the old and the new interface.

Reply:
There is a fair chance that there will be no overhead at all :) Looking forward to the numbers!